src/middlewared/middlewared/plugins/nfs.py

```diff
@@ -61,3 +64,233 @@ def nfs_compress(self, nfs):
+
+    @accepts(Int("id"))
+    async def do_delete(self, id):
+        await self.middleware.call("datastore.sql", "DELETE FROM sharing_nfs_share_path WHERE share_id = ?", (id,))
+        await self.middleware.call("datastore.delete", self._config.datastore, id)
+
+    @private
+    async def validate(self, data, schema_name, verrors, old=None):
+        if not data["paths"]:
+            verrors.add(f"{schema_name}.paths", "At least one path is required")
+
+        await self.middleware.run_in_io_thread(self.validate_paths, data, schema_name, verrors)
+
+        if not data["networks"]:
+            verrors.add(f"{schema_name}.networks", "At least one network is required")
+
+        for i, network1 in enumerate(data["networks"]):
+            network1 = ipaddress.ip_network(network1, strict=True)
+            for j, network2 in enumerate(data["networks"]):
+                if j > i:
+                    network2 = ipaddress.ip_network(network2, strict=True)
+                    if network1.overlaps(network2):
+                        verrors.add(f"{schema_name}.network.{j}", "Networks {network1} and {network2} overlap")
+
+        filters = []
+        if old:
+            filters.append(["id", "!=", old["id"]])
+        other_shares = await self.middleware.call("sharing.nfs.query", filters)
+        await self.middleware.run_in_io_thread(self.validate_user_networks, other_shares, data, schema_name, verrors)
+
+        for k in ["maproot", "mapall"]:
+            if not data[f"{k}_user"] and not data[f"{k}_group"]:
+                pass
+            elif not data[f"{k}_user"] and data[f"{k}_group"]:
+                verrors.add(f"{schema_name}.{k}_user", "This field is required when map group is specified")
+            elif data[f"{k}_user"] and not data[f"{k}_group"]:
+                verrors.add(f"{schema_name}.{k}_group", "This field is required when map user is specified")
+            else:
+                user = await self.middleware.call("user.query", [("username", "=", data[f"{k}_user"])])
+                if not user:
+                    verrors.add(f"{schema_name}.{k}_user", "User not found")
+
+                group = await self.middleware.call("group.query", [("group", "=", data[f"{k}_group"])])
+                if not group:
+                    verrors.add(f"{schema_name}.{k}_group", "Group not found")
+
+        if data["maproot_user"] and data["mapall_user"]:
+            verrors.add(f"{schema_name}.mapall_user", "maproot_user disqualifies mapall_user")
+
+        if data["security"]:
+            nfs_config = await self.middleware.call("nfs.config")
+            if not nfs_config["v4"]:
+                verrors.add(f"{schema_name}.security", "This is not allowed when NFS v4 is disabled")
+
+    @private
+    def validate_paths(self, data, schema_name, verrors):
+        dev = None
+        is_mountpoint = False
+        for i, path in enumerate(data["paths"]):
+            stat = os.stat(path.encode("utf8"))
```

Comment:
Does this need to be encoded? Last I knew it would encode automatically using system locale.